File: matlab/dlaed1.m
function [D, Q, perm1] = dlaed1(D, Q, perm1, rho, cutpnt)
% computes the updated eigensystem of a diagonal matrix after modification by a
% rank-one symmetric matrix.
%
%   T = Q(in) ( D(in) + rho * z * z') Q'(in) = Q(out) * D(out) * Q'(out)
%     where z = Q'(in) * u, u is a vector of length N with ones in the cutpnt
%     and cutpnt + 1 th elements and zeros elsewhere. 
N = length(D);

% form the z vector
z = [Q(cutpnt, 1:cutpnt)'; Q(cutpnt+1, cutpnt+1:end)']

% deflate eigenvalues
perm1(cutpnt+1:end) = perm1(cutpnt+1:end) + cutpnt;
[K, D, Q, rho, w] = dlaed2(D, Q, cutpnt, perm1, rho, z)

% solve secular equation
if K > 0
    [D(1:K), Qhat] = dlaed3(D(1:K), w, 1/rho);
    Qhat
    % back-transformation
    Q = [Q(:, 1:K) * Qhat, Q(:, K+1:end)];

    % compute perm1 that would merge back deflated values.
    perm1 = dlamrg(K, N - K, D, 1, -1);
else
    perm1 = 1:N;
end

File: c/dlaed1.c
#include <cblas.h>
#include <lapacke.h>
#include <stdio.h>
#include "dstedc.h"

void dlaed1(int N, double *D, double *Q, int LDQ, int *perm1, double RHO,
    int CUTPNT, double *WORK, int *IWORK)
/*
computes the updated eigensystem of a diagonal matrix after modification by a
rank-one symmetric matrix.

  T = Q(in) ( D(in) + rho * z * z') Q'(in) = Q(out) * D(out) * Q'(out)
    where z = Q'(in) * u, u is a vector of length N with ones in the cutpnt
    and cutpnt + 1 th elements and zeros elsewhere. 
*/
{
    int i;
    int K;
    double *Z      = &WORK[0];
    double *DWORK  = &WORK[N];
    double *QHAT   = &WORK[2 * N];
    double *QWORK  = &WORK[2 * N + N * N];

    int *perm2     = &IWORK[0];
    int *permacc   = &IWORK[N];
    int *perm3     = &IWORK[2 * N];

    // form the z vector
    cblas_dcopy(CUTPNT, &Q[CUTPNT-1], LDQ, &Z[0], 1);
    cblas_dcopy(N-CUTPNT, &Q[CUTPNT + CUTPNT * LDQ], LDQ,
        &Z[CUTPNT], 1);

    // deflate eigenvalues
    for (i = CUTPNT; i < N; i++)
        perm1[i] += CUTPNT;

    dlaed2(&K, N, CUTPNT, D, Q, LDQ, perm1, &RHO, Z, DWORK, QWORK, perm2,
        permacc, perm3);

    // sovle secular equation
    if (K > 0) {
        cblas_dcopy(K, D, 1, DWORK, 1);
        dlaed3(K, D, QHAT, K, RHO, DWORK, Z, QWORK);

        // back-transformation
        LAPACKE_dlacpy(LAPACK_COL_MAJOR, 'A', N, K, Q, LDQ, QWORK, N);
        cblas_dgemm(CblasColMajor, CblasNoTrans, CblasNoTrans, N, K, K,
            1.0, QWORK, N, QHAT, K, 0.0, Q, LDQ);

        // compute perm1 that would merge back deflated values.
        dlamrg(K, N-K, D, 1, -1, perm1);
    } else {
        for (i = 0; i < N; i++)
            perm1[i] = i;
    }
}
